framework/src/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -537,14 +539,26 @@ public Boolean visitTypeArgs(
                 return true;
             }
 
-            if (supertypeTypeArgs.size() > 0) {
-                for (int i = 0; i < supertypeTypeArgs.size(); i++) {
-                    final AnnotatedTypeMirror superTypeArg = supertypeTypeArgs.get(i);
-                    final AnnotatedTypeMirror subTypeArg = subtypeTypeArgs.get(i);
-                    if (!compareTypeArgs(
-                            subTypeArg, superTypeArg, supertypeRaw, subtypeRaw, visited)) {
-                        return false;
-                    }
+            final TypeElement supertypeElem =
+                    (TypeElement) supertype.getUnderlyingType().asElement();
+            List<Integer> covariantArgIndexes = null;
+            AnnotationMirror covam =
+                    supertype.atypeFactory.getDeclAnnotation(supertypeElem, Covariant.class);
+            if (covam != null) {
+                covariantArgIndexes =
+                        AnnotationUtils.getElementValueArray(covam, "value", Integer.class, false);
+            }
+
+            for (int i = 0; i < supertypeTypeArgs.size(); i++) {
+                final AnnotatedTypeMirror superTypeArg = supertypeTypeArgs.get(i);
+                final AnnotatedTypeMirror subTypeArg = subtypeTypeArgs.get(i);
+                final boolean covariant =
+                        this.covariantTypeArgs
```

Comment:
[Glacier](https://github.com/mcoblenz/Glacier/blob/master/src/edu/cmu/cs/glacier/GlacierTypeHierarchy.java#L21) uses it.  But it wouldn't be hard to adapt Glacier to override `compareTypeArgs`.

I agree that it's risky to allow covariance for all types and so the field should be removed.